### ISSUE TITLE
Understanding Nesting VJP Functions

### ISSUE THREAD

**Issue by theadamsabra** [google-research/last#30]
I have a question about `binary_op_broadcasting_test` specifically when it pertains to testing the `semirings.times` on Log semirings.

When looking at the [`expected` function nested in the test](https://github.com/google-research/last/blob/main/tests/semirings_test.py#L76-L79), we run the operation through `jax.vjp`. However, within `Log.times` there [`_logaddexp`](https://github.com/google-research/last/blob/main/last/semirings.py#L260-L280) which in itself is a custom vjp function. 

So if my understanding is correct, this means that we are converting a function (which contains a custom vjp) into another vjp for testing. 

First, is my understanding correct? 
Second, I wanted to understand how JAX handles something like this, because torch cannot allow a custom vjp function to be fed into the the torch equivalent of `jax.vjp`.

Thanks!

**Comment by kho** [comment]
> First, is my understanding correct?

Yes, we are asking for the VJP of a function part of which is computed with an operation that has a custom VJP implementation

**Comment by theadamsabra** [comment]
There are 2 different `vjp` functions in torch - I only used one in which I got a rather dubious (and not clear) error. I later found out through ad-hoc debugging that it comes from calling, for example, `torch.func.vjp(semiring.sum, x, y)` in which the specific `semiring.sum` has a custom `vjp` within it.

There is another `vjp` within the `autograd.functional` module that the base class for the custom `vjp` comes from. I have yet to test this and this perhaps could be more useful in investigating. I will test this today and see if I can write up the tests in a similar fashion to you. Otherwise, I will use something like `grad` as you mentioned. 

Thanks!
